server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,213 @@
+
+  override def handlerAdded(ctx: ChannelHandlerContext): Unit =
+    if (ctx.channel.isActive) {
+      initHandler(ctx)
+    }
+  override def channelActive(ctx: ChannelHandlerContext): Unit = {
+    channelGroup.add(ctx.channel)
+    initHandler(ctx)
+  }
+
+  private[this] def initHandler(ctx: ChannelHandlerContext): Unit = {
+    if (eventLoopContext == null)
+      eventLoopContext = ExecutionContext.fromExecutor(ctx.channel.eventLoop)
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    logger.error("Error while processing the request", cause)
+  }
+
+  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+    def replyWithError500(reason: Throwable): Unit = {
+      logger.error("Error while processing the request", reason)
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyWith503(): Unit = {
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.SERVICE_UNAVAILABLE)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyNotFound(): Unit = {
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NOT_FOUND)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res).close()
+    }
+
+    def replyWithRouteResponse(
+        req: HttpRequest,
+        serverResponse: ServerResponse[NettyResponse],
+        channelPromise: ChannelPromise,
+        byteBuf: ByteBuf
+    ): Unit = {
+      val res = new DefaultFullHttpResponse(req.protocolVersion(), HttpResponseStatus.valueOf(serverResponse.code.code), byteBuf)
+      res.setHeadersFrom(serverResponse, serverHeader)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, byteBuf.readableBytes())
+      val _ = ctx.writeAndFlush(res, channelPromise)
+    }
+
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        if (isShuttingDown.get()) {
+          logger.info("Rejecting WS handshake request because the server is shutting down.")
+          replyWith503()
+        } else {
+          ReferenceCountUtil.release(msg)
+          val (runningFuture, _) = unsafeRunAsync { () =>
+            route(NettyServerRequest(req.retain()))
+              .map {
+                case Some(response) => response
+                case None           => ServerResponse.notFound
+              }
+          }
+
+          runningFuture.onComplete {
+            case Success(serverResponse) if serverResponse == ServerResponse.notFound =>
```

Comment:
[minor] maybe we can simply .transform the original future (without the .map in :116), and if it's a none, return a 404